lectures/olg.md

```diff
@@ -46,114 +47,917 @@ The OLG model takes up this challenge.
+    rhs = L * (x / α)**(1/(α-1))
+    return lhs - rhs
+```
+
+```{code-cell} ipython3
+def find_Rstar(K_prev, model):
+    return optimize.newton(find_Rstar_newton, 0.5, args=(K_prev, model))
+
+def find_Kstar(R_star, model):
+    return model.L * (R_star / model.α)**(1/(model.α-1))
+```
+
+
+
+The following function plots the equilibrium quantity and equilibrium price.
+
+```{code-cell} ipython3
+def plot_ks_rs(K_t_vals, model):
+    n = len(K_t_vals)
+    R_star = np.empty(n)
+    K_star = np.empty(n)
+
+    for i in range(n):
+        R_star[i] = find_Rstar(K_t_vals[i], model)
+        K_star[i] = find_Kstar(R_star[i], model)
+
+    fig, ax = plt.subplots()
+
+    ax.plot(K_t_vals, R_star, label="equilibrium price")
+    ax.plot(K_t_vals, K_star, label="equilibrium quantity")
+
+    ax.set_xlabel("$K_{t}$")
+    ax.legend()
+    plt.show()
+```
+
+```{code-cell} ipython3
+---
+mystnb:
+  figure:
+    caption: 'Equilibrium price and quantity
+
+      '
+    name: equi_ps_q_crra
+  image:
+    alt: equi_ps_q_crra
+    classes: shadow bg-primary
+    width: 200px
+---
+K_t_vals = np.linspace(0.1, 50, 50)
+m_crra = create_olg_model(u=crra, u_params={'γ': 0.5})
+plot_ks_rs(K_t_vals, m_crra)
+```
+
+
+
+```{solution-end}
+```
+
+```{exercise}
+:label: olg_ex2
+
+Let's keep the model the same except for replacing the utility function $u(c)$ in {eq}`eq_crra`  with a nonlinear form $u(c)=c + c^{\theta}$.
+
+Like what we did in the CRRA case we don't have an analytical solution.
+
+Try to compute the time path capital $\{k_t\}$ in this case.
+```
+
+```{solution-start} olg_ex2
+:class: dropdown
+```
+
+To get the time path capital $\{k_t\}$ first we need to solve the household's utility maximization problem for the optimal consumption and optimal saving.
+
+With the quasilinear preference the Euler equation [](euler_2_olg) becomes
+
+```{math}
+:label: euler_quasilinear
+    1 + \theta (w_t - s_t)^{\theta-1} = \beta R_{t+1} + \beta R^{\theta}_{t+1} \theta s_t^{\theta - 1}
+```
+
+Let $k_t := K_t / L$.
+
+Since [](aggregate_supply_capital_log_olg), [](wage_2) and [](interest_rate_2) the Euler equation becomes
+
+```{math}
+:label: euler_quasilinear1
+    1 + \theta ((1-\alpha)k^{\alpha}_t - k_{t+1})^{\theta-1} = \beta \alpha k^{\alpha - 1}_{t+1} + \beta \theta \alpha^{\theta} k^{\alpha \theta - 1}_{t+1}
+```
+
+Obviously $k_{t+1}$ cannot be solved by pencil and paper.
+
+To solve for $k_{t+1}$ we need to turn to the newton's method.
+
+Let's start by defining the utility function.
+
+```{code-cell} ipython3
+def u_quasilinear(c, θ=3):
+    return c + c**θ
+```
+
+
+
+The function `find_k_next` is used to find $k_{t+1}$ by finding
+the root of equation [](euler_quasilinear1) using the helper
+function `solve_for_k_next` for a given value of $k_t$.
+
+```{code-cell} ipython3
+def solve_for_k_next(x, k_t, model):
+    α, β, L, θ = model.α, model.β, model.L, model.u_params['θ']
+    l = 1 + θ * ((1 - α) * k_t**α - x)**(θ - 1)
```

Comment:
Brute force does work fine, but the equation on line 897 gets trickier with fractional powers lie 1.5, 0.5, etc. because then we would have runtime warnings if the argument is < 0. Such cases are also difficult to handle on the grid unless we have an upper limit strictly defined as `(1 - α) * k_t**α - x` which sometimes might be unrealistic if $k_{t} < k^\*$

Reply:
Thanks @Smit-create .  What do you think @shlff ?

Reply:
Thanks @jstac and @Smit-create . 

I think it might be useful to plot the function $k_{t+1}$ as $k_t$. Please find the added figure at the end of the solution to exercise 2:
- https://6434b0a6db97f748a5148163--taupe-gaufre-c4e660.netlify.app/olg.html#olg_ex2
or here
<img width="1279" alt="Screenshot 2023-04-11 at 11 22 16 am" src="https://user-images.githubusercontent.com/53931041/231030909-3aa2a887-b302-43ae-b343-46851fccb9be.png">


If we zoom in, then we can see that
<img width="826" alt="Screenshot 2023-04-11 at 11 11 16 am" src="https://user-images.githubusercontent.com/53931041/231029626-7d22c8e6-8b14-47ba-821e-7e6c62335e31.png">

From the graph we can tell that this function $g:k_{t+1} \to k_t$ is not well-behaved (not concave).  Also, we have $k_{t+1}=k_t=k^*$ probably for infinite many $k^*$. 

Therefore I reckon one way to solve this is probably to have a 45-degree plot first and then consider the application of brute force or Newton's method, as well as the adoption of the initial value $k_0$.

Reply:
Thanks, @shlff. The plot looks interesting, and I guess for the same reason we are having this multiple $k^\*$ issue. This makes sense now. But still, we can see a runtime warning while plotting these graphs and I'm not sure how it is handled by scipy i.e does it give `nan` or `inf`.